community/kernel/src/main/java/org/neo4j/kernel/NeoStoreDataSource.java

```diff
@@ -548,25 +550,40 @@ public void transactionRecovered( long txId )
                     propertyKeyTokenHolder, labelTokens, relationshipTypeTokens,
                     indexingModule.indexingService(), cacheModule.schemaCache(), cacheModule.procedureCache() );
 
+            SafeFreeIdManagement freeIdManagement = dependencies.satisfyDependency( new SafeFreeIdManagement(
+                    neoStoreModule.neoStores(), scheduler ) ); // <-- deferred KernelTransactions dependency
+            life.add( freeIdManagement );
             TransactionLogModule transactionLogModule =
                     buildTransactionLogs( storeDir, config, logProvider, scheduler, indexingModule.labelScanStore(),
                             fs, neoStoreModule.neoStores(), cacheModule.cacheAccess(), indexingModule.indexingService(),
-                            indexProviders.values(), legacyIndexApplierLookup );
+                            indexProviders.values(), legacyIndexApplierLookup, freeIdManagement );
 
             buildRecovery( fs, cacheModule.cacheAccess(), indexingModule.indexingService(),
                     indexingModule.labelScanStore(), neoStoreModule.neoStores(),
                     monitors.newMonitor( RecoveryVisitor.Monitor.class ), monitors.newMonitor( Recovery.Monitor.class ),
                     transactionLogModule.logFiles(), transactionLogModule.storeFlusher(), startupStatistics,
                     legacyIndexApplierLookup );
 
-            KernelModule kernelModule = buildKernel( indexingModule.integrityValidator(),
+            final KernelModule kernelModule = buildKernel( indexingModule.integrityValidator(),
                     transactionLogModule.transactionAppender(), neoStoreModule.neoStores(),
                     transactionLogModule.storeApplier(), indexingModule.indexingService(),
                     indexingModule.indexUpdatesValidator(),
                     storeLayerModule.storeLayer(),
                     cacheModule.updateableSchemaState(), indexingModule.labelScanStore(),
                     indexingModule.schemaIndexProviderMap(), cacheModule.procedureCache() );
 
+            // This oddity is here to break this circle:
+            // KernelTransactions -> TransactionCommitProcess -> TransactionRepresentationStoreApplier -> SafeFreeIdManagement
+            //         ^-----------------------------------------------------------------------------------------------|
+            freeIdManagement.initialize( new Supplier<KernelTransactions.Snapshot>()
+            {
+                @Override
+                public KernelTransactions.Snapshot get()
+                {
+                    return kernelModule.kernelTransactions().snapshot();
+                }
+            } );
+
```

Comment:
I'm not happy with this entangling. Do you plan to redo the design of this in 3.0 or 3.1?